x-pack/plugins/security/server/authentication/authentication_service.ts

```diff
@@ -139,7 +157,41 @@ export class AuthenticationService {
       return t.notHandled();
     });
 
-    this.logger.debug('Successfully registered core authentication handler.');
+    http.registerOnPreResponse((request, preResponse, toolkit) => {
+      if (preResponse.statusCode !== 401 || !canRedirectRequest(request)) {
+        return toolkit.next();
+      }
+
+      if (!this.authenticator) {
+        // Core doesn't allow returning error here.
+        this.logger.error('Authentication sub-system is not fully initialized yet.');
+        return toolkit.next();
+      }
+
+      // If users can eventually re-login we want to redirect them directly to the page they tried
+      // to access initially, but we only want to do that for non-API routes. API routes that support
+      // redirects are solely used to support various authentication flows that wouldn't make any
+      // sense after successful authentication through login page.
+      const urlToRedirectToAfterLogin = !API_ROUTES_SUPPORTING_REDIRECTS.includes(
+        request.route.path
+      )
+        ? `${NEXT_URL_QUERY_STRING_PARAMETER}=${encodeURIComponent(
+            this.authenticator.getRequestOriginalURL(request)
+          )}`
+        : '';
+
+      return toolkit.render({
+        body: `
+          <html lang=${i18n.getLocale()}>
+            <head>
+              <title>Unauthorized</title>
+              <link rel="icon" href="data:," />
+              <meta http-equiv="refresh" content="0;url=${unauthorizedURL}?${urlToRedirectToAfterLogin}&msg=UNAUTHORIZED" />
```

Comment:
I've considered three different options here:

* Use the same approach as we used for 403 interceptor and statically render `Unauthorized` page. It would be hacky, but it's tolerable. The main reason why I ruled it out is that I believe we should render login page/selector if it's available since trying to log in again is the only action user can take in this case anyway.

* Redirect directly from here to a proper page. There are three reasons why I didn't like it:
  * Core doesn't allow us to redirect from `onPreResponse` yet, it shouldn't be difficult to add though
  * I believe it's just more correct to return 401 in this case, and we'd also avoid breaking any users' tooling that relies on 401 code for whatever reason
  * We'll have to specifically handle Kerberos (or anything that returns `WWW-Authenticate`), assuming core will allow us to access response headers

* Return 401 with the HTML body containing `http-equiv="refresh"`. I cannot say I'm happy about this solution, but it seems to give us everything we need - both 401 for tooling and Kerberos and automatic redirect for browsers.

Reply:
I appreciate all of the consideration you put into the different approaches, and thanks for describing your thought process here.

I agree with you about option 2: issuing a redirect to the proper page is my least favorite option.

What do you think about combining options 1 and 3? We could statically render the `Unauthorized` page, but include a simple `setTimeout()` script that performs the redirect if we detect that the user should be taken to the login page/selector.